File: HongJun/Tasks/Libraries/ODE/testScript/testScript.m

function main()
    %This script is used for the random testing.
    %Below is an example of the possible use of the random testing.
    
    %Function to test, is not randomised
    fun = @cosbasic;
    
    %Size of the system
    size = 2;
    
    %Interval for the tspan
    tstart=-100;
    tend=100;
    
    %Interval for initial values
    y0start=1;
    y0end=100;
    
    %Event function
    events=@eventsbasic;   
    
    %Mass function, or mass matrix 
    mass=[];
    
    % Choice of the options to test, any combination of the following 
    % choices, (Note: 'massmatrix','massfunctime', and 'massfuncstate' 
    % should not be used at the same time):
    %
    %   'tol' - Will randomise AbsTol, RelTol, and NormControl
    %   'nonnegative' - Will randomise NonNegative
    %   'step' - Will randomise MaxStep and InitialStep
    %   'refine' - Will randomise Refine
    %   'events' - Will use the event function passed
    %   'massmatrix' - Will the mass as a matrix
    %   'massfunctime' - Will use mass as a time depedent function
    %   'massfuncstate' - Will use mass as a state-time depedent function
    %
    choices={'nonnegative','tol','step','refine','massmatrix'};
    fileID = fopen('test.txt','w');
    
    %Loop to execute a random test, in this example 10 tests will be
    %performed
    for i=1:10
        execute_test(fun,size,tstart,tend,y0start,y0end,events,mass,choices,fileID);
    end
    
    
end

function execute_test(fun,size,tstart,tend,y0start,y0end,events,mass,choices,textfile)
    succ = false;
    while ~succ
        [y0,tspan]=randInput(size,tstart,tend,y0start,y0end);
        options = odeset();
        te=[];
        ye=[];
        ie=[];
        if any(strcmp(choices,'tol'))
            options = randTol(options,y0);
        end
        if any(strcmp(choices,'refine'))
            options = randRefine(options);
        end
        if any(strcmp(choices,'nonnegative'))
            options = randNN(options,y0);
        end
        if any(strcmp(choices,'step'))
            options = randStep(options);
        end
        if any(strcmp(choices,'events'))
            options = odeset(options,'Events',events);
        end
        if any(strcmp(choices,'massmatrix'))
            options = randMatrix(options,y0);
        end
        if any(strcmp(choices,'massfunctime'))
            options = odeset(options,'Events',mass,'MStateDependence','none');
        end
        if any(strcmp(choices,'massfuncstate'))
            options = odeset(options,'Mass',mass,'MStateDependence','weak');
        end
        try
            if any(strcmp(choices,'events'))
                [t,y,te,ye,ie]=ode45(fun,tspan,y0,options);
            else
                [t,y]=ode45(fun,tspan,y0,options);
            end
            sol=ode45(fun,tspan,y0,options);
            succ = true;
        catch
            warning('Test failed');
        end
    end
    writetext(textfile,sol,tspan,y0,t,y,te,ye,ie);
end



%%
%
%   Write your ode, event, mass functions here
%
%

function dydt = cosbasic(t,y)
    dydt=[cos(t);2*cos(t)];
end

function dydt = trigbasic(t,y)
    dydt = [y(1)*cos(t);y(2)*sin(t)];
end

function dydt = polybasic(t,y)
    dydt = [0.02*(3*t^5-52*t^3+42*t^2+145*t+18);0.02*(3*t^5-52*t^3+42*t^2+145*t+18)];
end

function [value,isterminal,direction] = eventsbasic(t,y)
    value = [y(1),y(2)];
    isterminal = [1,0];
    direction = [1,-1];
end

%%



%%
function [y0,tspan] = randInput(size,start,en,y0start,y0end)
    bool = [0,1];
    y0exponent = zeros(1,floor(log10(y0end/y0start))+1);
    y0exponent(1) = y0start;
    for i = 2:length(y0exponent)
        y0exponent(i) = y0exponent(i-1)*10;
    end
    y0=zeros(size,1);
    for i=1:size
        y0(i) = rand * randsample(y0exponent,1);
    end
    if randsample(bool,1)
        t0= start + (en-start) .* rand;
        tend = t0 +(en-t0) .* rand;
        tspan = [t0,tend];
    elseif randsample(bool,1)
        t0= start + (en-start) .* rand;
        tend = t0 +(en-t0) .* rand;
        space = 1/ (50 + (1000-50) *rand);
        tspan = [t0:space:tend];
    else
        t0= start + (en-start) .* rand;
        tend = t0 +(en-t0) .* rand;
        space = 1 / (50 + (1000-50) *rand);
        num = round((tend-t0)/space);
        tspan=sort(rand(1,num)*(tend-t0) + t0);
        
    end
end

function option = randTol(opt,y0)
    bool = [0,1];
    absexponent = [1e-10,1e-9,1e-8,1e-7,1e-6,1e-5,1e-4,1e-3];
    relexponent = [1e-8,1e-7,1e-6,1e-5,1e-4,1e-3,1e-2,1e-1,1];
    randNorm=randsample(bool,1);
    if randNorm
        option=odeset(opt,'NormControl','on');
    else
        option=odeset(opt,'NormControl','off');
    end
    absArray=randsample(bool,1);
    if absArray && ~randNorm
        randAbs=rand(1,length(y0)).*randsample(absexponent,1);
    else
        randAbs=rand*randsample(absexponent,1);
    end
    option=odeset(option,'AbsTol',randAbs);
    option=odeset(option,'RelTol',rand*randsample(relexponent,1));
end

function option = randRefine(opt)
    option = odeset(opt,'Refine',floor(1 + 20 * rand));
end

function option = randNN(opt,y0)
    bool = [0,1];
    size = length(y0);
    index = [];
    for i = 1:size
        if randsample(bool,1)
            index = [index, i];
        end
    end
    option = odeset(opt,'NonNegative',index);
end

function option = randStep(opt)
    maxexponent = [1e-3,1e-2,1e-1,1,10];
    initexponent = [1e-5,1e-4,1e-3,1e-2,1e-1,1];
    option=odeset(opt,'MaxStep',rand*randsample(maxexponent,1));
    option=odeset(option,'InitialStep',rand*randsample(initexponent,1));
end


function option = randMatrix(opt,y0)
    massexponent = [1e-1,1,1e1];
    size = length(y0);
    mat = zeros(size,size);
    for i = 1:size
        for j = 1:size
            mat(i,j)=rand*randsample(massexponent,1);
        end
    end
    option = odeset(opt,'Mass',mat);
end


%%

%%

function writetext(fileID,sol,tspan,y0,t,y,te,ye,ie)
    
    %Inputs
    fprintf(fileID,'Function:');
    fprintf(fileID,'%s',func2str(sol.extdata.odefun));
    fprintf(fileID,' ');
    fprintf(fileID,'Tspan:');
    fprintf(fileID,'%.15f#',tspan);
    fprintf(fileID,' ');
    fprintf(fileID,'Y0:');
    fprintf(fileID,'%.15f#',y0);
    fprintf(fileID,' ');
    fprintf(fileID,'Varargin:');
    fprintf(fileID,'%.15f#',string(sol.extdata.varargin));
    fprintf(fileID,' ');
    
    %Options
    fprintf(fileID,'RelTol:');
    fprintf(fileID,'%.30f',sol.extdata.options.RelTol);
    fprintf(fileID,' ');
    fprintf(fileID,'AbsTol:');
    fprintf(fileID,'%.30f#',sol.extdata.options.AbsTol);
    fprintf(fileID,' ');
    fprintf(fileID,'NormControl:');
    fprintf(fileID,'%s',sol.extdata.options.NormControl);
    fprintf(fileID,' ');
    fprintf(fileID,'Refine:');
    fprintf(fileID,'%d',sol.extdata.options.Refine);
    fprintf(fileID,' ');
    fprintf(fileID,'Stats:');
    fprintf(fileID,'%s',sol.extdata.options.Stats);
    fprintf(fileID,' ');
    fprintf(fileID,'NonNegative:');
    fprintf(fileID,'%d#',sol.extdata.options.NonNegative);
    fprintf(fileID,' ');
    fprintf(fileID,'Events:');
    if isa(sol.extdata.options.Events,'function_handle')
        fprintf(fileID,'%s',func2str(sol.extdata.options.Events));
    end
    fprintf(fileID,' ');
    fprintf(fileID,'MaxStep:');
    fprintf(fileID,'%.15f',sol.extdata.options.MaxStep);
    fprintf(fileID,' ');
    fprintf(fileID,'InitialStep:');
    fprintf(fileID,'%.15f',sol.extdata.options.InitialStep);
    fprintf(fileID,' ');
    fprintf(fileID,'Mass:');
    if isa(sol.extdata.options.Mass,'function_handle')
        fprintf(fileID,'%s',func2str(sol.extdata.options.Mass));
    else
        fprintf(fileID,'%.15f#',sol.extdata.options.Mass);
    end
    fprintf(fileID,' ');
    fprintf(fileID,'MStateDependence:');
    fprintf(fileID,'%s',sol.extdata.options.MStateDependence);
    fprintf(fileID,' ');

    %Output
    fprintf(fileID,'Tout:');
    fprintf(fileID,'%.15f#',t);
    fprintf(fileID,' ');
    fprintf(fileID,'Yout:');
    fprintf(fileID,'%.15f#',y);
    fprintf(fileID,' ');
    fprintf(fileID,'Nsteps:');
    fprintf(fileID,'%d',sol.stats.nsteps);
    fprintf(fileID,' ');
    fprintf(fileID,'Nfailed:');
    fprintf(fileID,'%d',sol.stats.nfailed);
    fprintf(fileID,' ');
    fprintf(fileID,'Nfevals:');
    fprintf(fileID,'%d',sol.stats.nfevals);
    fprintf(fileID,' ');
    fprintf(fileID,'Teout:');
    if isa(sol.extdata.options.Events,'function_handle')
        fprintf(fileID,'%.15f#',te);
    end
    fprintf(fileID,' ');
    fprintf(fileID,'Yeout:');
    if isa(sol.extdata.options.Events,'function_handle')
        fprintf(fileID,'%.15f#',ye);
    end
    fprintf(fileID,' ');
    fprintf(fileID,'Ieout:');
    if isa(sol.extdata.options.Events,'function_handle')
        fprintf(fileID,'%d#',ie);
    end
    fprintf(fileID,'\n');
    
end

File: HongJun/Tasks/Libraries/ODE/testScript/testScript.py
import numpy as np
import os
import sys

currentdir = os.path.dirname(os.path.realpath(__file__))
parentdir = os.path.dirname(currentdir)
sys.path.append(parentdir)

from ode import ode45

'''The This script is used to perform random testing for the ode45 function. This script
    works in conjunction with testScript.m, which generates the expected output for a series 
    of random inputs. This script will then read the inputs and expected outputs, pass the 
    inputs to the implemented ode45 function and check them against the expected output.
    Note: any functions which are added to the MATLAB script also have to be added to this 
    script.
'''

def solve_ode(inputs,result):
    sol=ode45(inputs.fun,inputs.tspan,inputs.y0,inputs.get_options(),inputs.varargin)
    result.compare_ty(sol.get_t(),sol.get_y())
    result.compare_stats(sol.get_stats())


def get_fun(val):
    #Add any ode function here
    f=None
    if val == 'cosbasic':
        def f(t,y):
            return [np.cos(t),2*np.cos(t)]
    if val == 'trigbasic':
        def f(t,y):
            return [y[0]*np.cos(t),y[1]*np.sin(t)]
    return f

def get_event_fun(val):
    #Add any event function here
    return val

def get_mass_fun(val):
    #Add any mass function here
    return val

class Inputs:
    def __init__(self):
        self.fun=None
        self.tspan=None
        self.y0=None
        self.varargin=[]
        self.reltol=None
        self.norm=None
        self.abstol=None
        self.refine=None
        self.stats=None
        self.nonnegative=None
        self.events=None
        self.maxstep=None
        self.initialstep=None
        self.mass=None
        self.massstate=None

    def __str__(self):
        if len(self.tspan)>2:
            strtspan=str(len(self.tspan))
        else:
            strtspan=str(self.tspan)
        return "Fun : "+str(self.fun)+"\nTspan : "+strtspan+"\nY0 : "+str(self.y0)+ \
        "\nVarargin : "+str(self.varargin)+"\nRelTol : "+str(self.reltol)+"\nAbsTol : "+str(self.abstol)+ \
        "\nNormControl : "+str(self.norm)+"\nRefine : "+str(self.refine)+"\nStats : "+str(self.stats)+ \
        "\nNonNegative : "+str(self.nonnegative)+"\nEvents : "+str(self.events)+"\nMaxStep : "+str(self.maxstep)+ \
        "\nInitialStep : "+str(self.initialstep)+"\nMass : "+str(self.mass)+"\nMStateDependence : "+str(self.massstate)+ "\n\n"

    def set_fun(self, val):
        self.fun=get_fun(val)
        
    def set_tspan(self, val):
        x = [0] * len(val)
        for i in range(len(val)):
            x[i]=float(val[i])
        self.tspan=x
    
    def set_y0(self, val):
        x = [0] * len(val)
        for i in range(len(val)):
            x[i]=float(val[i])
        self.y0=x
    
    def set_varargin(self, val):
        x = [0] * len(val)
        for i in range(len(val)):
            x[i]=float(val[i])
        self.varargin=x
    
    def set_reltol(self, val):
        self.reltol=float(val)
        
    def set_abstol(self, val):
        if len(val)==1:
            self.abstol=float(val[0])
        else:
            x = [0] * len(val)
            for i in range(len(val)):
                x[i]=float(val[i])
            self.abstol=x
        
    def set_norm(self, val):
        self.norm=val
        
    def set_refine(self, val):
        self.refine=int(val)
        
    def set_stats(self, val):
        self.stats=val
        
    def set_nonnegative(self, val):
        x = [0] * len(val)
        for i in range(len(val)):
            x[i]=int(val[i])-1
        self.nonnegative=x
        
    def set_events(self, val):
        self.events=get_event_fun(val)
        
    def set_maxstep(self, val):
        self.maxstep=float(val)
        
    def set_initialstep(self, val):
        self.initialstep=float(val)
    
    def set_mass_fun(self, val):
        self.mass=get_mass_fun(val)
    
    def set_mass_mat(self, val):
        l = int(np.sqrt(len(val)))
        x = np.zeros((l,l))
        for i in range(l):
            for j in range(l):
                x[i,j]=float(val[j*l + i])
        self.mass=x
    
    def set_massstate(self, val):
        self.massstate=val
    
    
    def get_options(self):
        opt ={}
        
        if self.reltol != None:
            opt["RelTol"] = self.reltol
        if self.abstol != None:
            opt["AbsTol"] = self.abstol
        if self.norm != None:
            opt["NormControl"] = self.norm
        if self.refine != None:
            opt["Refine"] = self.refine
        if self.stats != None:
            opt["Stats"] = self.stats
        if self.nonnegative != None:
            opt["NonNegative"] = self.nonnegative
        if self.events != None:
            opt["Events"] = self.events
        if self.maxstep != None:
            opt["MaxStep"] = self.maxstep
        if self.initialstep != None:
            opt["InitialStep"] = self.initialstep
        if not isinstance(self.mass,type(None)):
            opt["Mass"] = self.mass
        if self.massstate != None:
            opt["MStateDependence"] = self.massstate
        
        return opt

class Results:
    def __init__(self):
        self.tout=np.array([])
        self.yout=np.array([])
        self.nsteps=0
        self.nfailed=0
        self.nfevals=0
        self.teout=np.array([])
        self.yeout=np.array([])
        self.ieout=np.array([])
    
    
    def __str__(self):
        return "NStesps : "+str(self.nsteps)+"\nNFailed : "+str(self.nfailed)+"\nNFevals : "+str(self.nfevals)+ \
        "\nTout : "+str(self.tout)+"\nYout : "+str(self.yout)+"\nTeout : "+str(self.teout)+ \
        "\nYeout : "+str(self.yeout)+"\nIeout : "+str(self.ieout)+"\n\n"
    
    def compare_ty(self,tout,yout):
        errt = 0.0
        erry = 0.0
        
        t = self.tout
        y = self.yout
        
        print("Equal size : " + str(len(t)==len(tout) and len(y)==len(yout)*len(yout[0])) + " |t| = "+str(len(t))+" |y| = "+str(yout.shape))
        
        for i in range(len(t)):
            for j in range(len(yout)):
                if abs(yout[j][i]-y[j*len(t)+i])>erry:
                    erry = abs(yout[j][i]-y[j*len(t)+i])
            if abs(tout[i]-t[i])>errt:
                errt = abs(tout[i]-t[i])
        
        print("Error t : " + str(errt))
        print("Error y : " + str(erry))
        print()
    
    def compare_stats(self,statsvec):
        print("Nsteps Expected : "+str(self.nsteps)+" Actual : "+str(statsvec[0]))
        print("Nfailed Expected : "+str(self.nfailed)+" Actual : "+str(statsvec[1]))
        print("Nfevals Expected : "+str(self.nfevals)+" Actual : "+str(statsvec[2]))
        print()
        
    

        
    def compare_events(self,teout,yeout,ieout):
        te=self.teout
        ye=self.yeout
        ie=self.ieout
        
        
        print("Equal event size : " + str(len(self.teout)==len(teout)))
        print("Equal event size : " + str(len(self.teout)==len(teout)))
        print("Equal event size : " + str(len(self.teout)==len(teout)))
        self.assertEqual(len(te),len(teout))
        self.assertEqual(len(ye),len(yeout)*len(yeout[0]))
        self.assertEqual(len(ie),len(ieout))
        
        for i in range(len(te)):
            for j in range(len(yeout)):
                self.assertAlmostEqual(ye[j*len(te)+i],yeout[j][i])
            self.assertAlmostEqual(teout[i],te[i])
            self.assertEqual(ieout[i],ie[i]-1)
    
    
    def set_statvec(self, statsvec):
        self.nsteps=int(statsvec[0])
        self.nfailed=int(statsvec[1])
        self.nfevals=int(statsvec[2])
    
    def set_tout(self, tout):
        x = np.zeros((len(tout),))
        for i in range(len(tout)):
            x[i]=float(tout[i])
        self.tout=x
    
    def set_yout(self, yout):
        x = np.zeros((len(yout),))
        for i in range(len(yout)):
            x[i]=float(yout[i])
        self.yout=x
    
    def set_teout(self, teout):
        x = np.zeros((len(teout),))
        for i in range(len(teout)):
            x[i]=float(teout[i])
        self.teout=x
    
    def set_yeout(self, yeout):
        x = np.zeros((len(yeout),))
        for i in range(len(yeout)):
            x[i]=float(yeout[i])
        self.yeout=x

    def set_ieout(self, ieout):
        x = np.zeros((len(ieout),))
        for i in range(len(ieout)):
            x[i]=int(ieout[i])
        self.ieout=x
        
        
def read_tests(filename,names):
    f=open(filename,"r")
    results = []
    inputs = []
    for line in f:
        statsvec=[0,0,0]
        result = Results()
        inp = Inputs()
        sp = line.split(" ")
        for word in sp:
            key, values =word.split(":")
            if key == "Function":
                if not names:
                    inp.set_fun(values)
                else:
                    inp.fun=values
            elif key == "Tspan":
                val = values.split("#")
                val.pop()
                inp.set_tspan(val)         
            elif key == "Y0":
                val = values.split("#")
                val.pop()
                inp.set_y0(val) 
            elif key == "Varargin":
                val = values.split("#")
                val.pop()
                if val[0]!='':
                    inp.set_varargin(val)
            elif key == "RelTol":
                if values != '': 
                    inp.set_reltol(values)
            elif key == "AbsTol":
                val = values.split("#")
                val.pop()
                if val[0] != '': 
                    inp.set_abstol(val)
            elif key == "NormControl":
                if values != '': 
                    inp.set_norm(values)
            elif key == "Refine":
                if values != '': 
                    inp.set_refine(values)
            elif key == "Stats":
                if values != '': 
                    inp.set_stats(values)
            elif key == "NonNegative":
                val = values.split("#")
                val.pop()
                if val[0] != '':
                    inp.set_nonnegative(val)
            elif key == "Events":
                if values != '': 
                    if not names:
                        inp.set_events(values)
                    else:
                        inp.events=values
            elif key == "MaxStep":
                if values != '': 
                    inp.set_maxstep(values)
            elif key == "InitialStep":
                if values != '': 
                    inp.set_initialstep(values)
            elif key == "Mass":
                val = values.split("#")
                if len(val) == 1:
                    if not names:
                        inp.set_mass_fun(values)
                    else:
                        inp.mass=values
                else:
                    val.pop()
                    if val[0]!='':
                        inp.set_mass_mat(val)
            elif key == "MStateDependence":
                if values != '': 
                    inp.set_massstate(values)
            elif key == "Tout":
                val = values.split("#")
                val.pop()
                if len(val)>=1:
                    result.set_tout(val)
            elif key == "Yout":
                val = values.split("#")
                val.pop()
                if len(val)>=1:
                    result.set_yout(val)
            elif key == "Nsteps":
                statsvec[0]=values
            elif key == "Nfailed":
                statsvec[1]=values
            elif key == "Nfevals":
                statsvec[2]=values
            elif key == "Teout":
                val = values.split("#")
                val.pop()
                if len(val)>=1:
                    result.set_teout(val)
            elif key == "Yeout":
                val = values.split("#")
                val.pop()
                if len(val)>=1:
                    result.set_yeout(val)
            elif key == "Ieout":
                val = values.split("#")
                val.pop()
                if len(val)>=1:
                    result.set_ieout(val)
            else:
                print(key)
                raise Exception("Invalid key")
            
        result.set_statvec(statsvec)
        results.append(result)
        inputs.append(inp)
    return results,inputs

if __name__ == "__main__":
    results,inputs=read_tests("test.txt",False)
    for i in range(len(inputs)):
        solve_ode(inputs[i],results[i])

    